x/quarantine/keys_test.go

```diff
@@ -0,0 +1,1135 @@
+			key:      makeKey(quarantine.OptInPrefix, len(testAddr0)+1, testAddr0),
+			expPanic: fmt.Sprintf("expected key of length at least %d, got %d", len(testAddr0)+1+2, len(testAddr0)+2),
+		},
+		{
+			name:     "from CreateOptInKey addr 0",
+			key:      quarantine.CreateOptInKey(testAddr0),
+			expected: testAddr0,
+		},
+		{
+			name:     "from CreateOptInKey addr 1",
+			key:      quarantine.CreateOptInKey(testAddr1),
+			expected: testAddr1,
+		},
+		{
+			name:     "from CreateOptInKey addr 2",
+			key:      quarantine.CreateOptInKey(testAddr2),
+			expected: testAddr2,
+		},
+		{
+			name:     "from CreateOptInKey longAddr",
+			key:      quarantine.CreateOptInKey(longAddr),
+			expected: longAddr,
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			var actual sdk.AccAddress
+			testFunc := func() {
+				actual = quarantine.ParseOptInKey(tc.key)
+			}
+			if len(tc.expPanic) == 0 {
+				if assert.NotPanics(t, testFunc, "ParseOptInKey") {
+					assert.Equal(t, tc.expected, actual, "ParseOptInKey result")
+				}
+			} else {
+				assert.PanicsWithValue(t, tc.expPanic, testFunc, "ParseOptInKey")
+			}
+		})
+	}
+}
+
+func TestCreateAutoResponseToAddrPrefix(t *testing.T) {
+	expectedPrefix := quarantine.AutoResponsePrefix
+	testAddr0 := testutil.MakeTestAddr("cartap", 0)
+	testAddr1 := testutil.MakeTestAddr("cartap", 1)
+	badAddr := testutil.MakeBadAddr("cartap", 2)
+
+	t.Run("starts with AutoResponsePrefix", func(t *testing.T) {
+		key := quarantine.CreateAutoResponseToAddrPrefix(testAddr0)
+		actual := key[:len(expectedPrefix)]
+		assert.Equal(t, expectedPrefix, actual, "key prefix")
+	})
+
+	makeExpected := func(addrBz []byte) []byte {
+		rv := make([]byte, 0, len(expectedPrefix)+1+len(addrBz))
+		rv = append(rv, expectedPrefix...)
+		rv = append(rv, byte(len(addrBz)))
+		rv = append(rv, addrBz...)
+		return rv
+	}
+
+	tests := []struct {
+		name     string
+		toAddr   sdk.AccAddress
+		expected []byte
+		expPanic string
+	}{
+		{
+			name:     "addr 0",
+			toAddr:   testAddr0,
+			expected: makeExpected(testAddr0),
+		},
+		{
+			name:     "addr 1",
+			toAddr:   testAddr1,
+			expected: makeExpected(testAddr1),
+		},
+		{
+			name:     "nil",
+			toAddr:   nil,
+			expected: expectedPrefix,
+		},
+		{
+			name:     "too long",
+			toAddr:   badAddr,
+			expected: nil,
+			expPanic: fmt.Sprintf("address length should be max %d bytes, got %d: unknown address", address.MaxAddrLen, len(badAddr)),
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			var actual []byte
+			testFunc := func() {
+				actual = quarantine.CreateAutoResponseToAddrPrefix(tc.toAddr)
+			}
+			if len(tc.expPanic) == 0 {
+				if assert.NotPanics(t, testFunc, "CreateAutoResponseToAddrPrefix") {
+					assert.Equal(t, tc.expected, actual, "CreateAutoResponseToAddrPrefix result")
+				}
+			} else {
+				assert.PanicsWithError(t, tc.expPanic, testFunc, "CreateAutoResponseToAddrPrefix")
+			}
+		})
+	}
+}
```

Comment:
The `CreateAutoResponseToAddrPrefix` function calls `MustLengthPrefix`, which is expected to panic in some situations. That means that we expect `CreateAutoResponseToAddrPrefix` to panic sometimes too. So I wanted a test case that panicked on that line. The only way to make sure that the panic is coming from the expected line is to assert that the panic message is the one from that `MustLengthPrefix`. If I only assert that `CreateAutoResponseToAddrPrefix` panics, but don't check the panic message, the test case might erroneously pass because something unexpected is causing the panic instead.